### PR TITLE
Add sgd docs

### LINKED ISSUE #4527: [Term Entry] Neural Networks ai stochastic gradient descent
### Reason/inspiration (optional)

"We would like a new term entry in the `AI` concept for [neural-network](https://www.codecademy.com/resources/docs/ai/neural-networks):  stochastic gradient descent. The entry should go in a new file under `docs/content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md`.

The entry should describe the term and how it is used. Also include: 
- An "Example" section that explains to readers how to use concept for stochastic gradient descent.
- Should contain information for this concept and syntax.
- Should have a Python codebyte with a working code snippet showcasing this concept.
Please refer to the [entry templates](https://github.com/Codecademy/docs/blob/main/documentation/), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue."

### Entry Type (select all that apply)

New Term for Existing Concept Entry

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by pratik305**
### Description

This pull request introduces a new documentation file on **Stochastic Gradient Descent (SGD)** in the `stochastic-gradient-descent.md` file. 
SGD is a widely used optimization algorithm in machine learning and deep learning due to its efficiency and scalability, particularly for large datasets. However, its effectiveness and stability can be significantly influenced by how it is implemented and tuned. This new documentation aims to offer clear, comprehensive insights into SGD’s operation, benefits, and limitations, helping users to better understand and apply this algorithm in their projects.

### Issue Solved

Closes #4527 

### Type of Change

- [x] Adding a new entry
- Editing an existing entry (fixing a typo, bug, issues, etc)
- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

**Comment by pratik305** [comment]
@SaviDahegaonkar  i am facing issue related to workflow awaiting approval how can i solve it

**Comment by pratik305** [comment]
hi @SaviDahegaonkar Could you please review this PR when you get a chance? Let me know if any adjustments are needed.
Thanks

**Comment by pratik305** [comment]
hi @SaviDahegaonkar Could you please review this PR when you get a chance? Let me know if any adjustments are needed.
Thanks

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+Title: 'Stochastic Gradient Desent'
+Description: 'Stochastic Gradient Desent is optimizer algorithm that minimizes the loss functions in machine learning and deep learning models.'
+Subjects:
+  - 'Machine Learning'
+  - 'Deep Learning'
+  - 'Computer Science'
+Tags:
+  - 'AI'
+  - 'Neural Network'
+  - 'Optimizer'
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:11]:
```suggestion
  - 'AI'
  - 'Neural Network'
  - 'Optimizer'
```
Include only those tags that are part of [tags.md](documentation/tags.md) list. Here `optimizer` is not part of the tags list so you can include it in the tags.md list so you can use this in your PR.

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+CatalogContent:
+  - 'paths/computer-science'
+  - 'paths/data-science'
+---
+
+**Stochastic Gradient Descent** (SGD) is a optimization algorithm. It is variant of gradient descent optimizer. The SGD minimize the loss function of machine learning algorithms and deep learning algorithms during backpropagation to update the weight and bias in Artificial Neural Networks.
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:17]:
```suggestion
**Stochastic Gradient Descent** (SGD) is an optimization algorithm. It is variant of gradient descent optimizer. The SGD minimizes the loss function of machine learning algorithms and deep learning algorithms during backpropagation to update the weights and biases in Artificial Neural Networks. 
```

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  - 'Neural Network'
+  - 'Optimizer'
+CatalogContent:
+  - 'paths/computer-science'
+  - 'paths/data-science'
+---
+
+**Stochastic Gradient Descent** (SGD) is a optimization algorithm. It is variant of gradient descent optimizer. The SGD minimize the loss function of machine learning algorithms and deep learning algorithms during backpropagation to update the weight and bias in Artificial Neural Networks. 
+
+The term stochastic mean randomness on which algorithm based upon. In this algorithm instead of taking whole dataset like grdient descent we take single randomly selected data point or small batch of data.suppose if the data set contains 500 rows SGD update the model parameters 500 times in one cycle or one epoch.
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:19]:
```suggestion
The term `stochastic` means randomness on which the algorithm is based. In this algorithm, instead of taking whole datasets like `gradient descent`, we take single randomly selected data points or small batches of data. Suppose if the data set contains 500 rows SGD updates the model parameters 500 times in one cycle or one epoch.
```

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+- At each iteration, a random sample  is selected from the training dataset.
+- The gradient of the cost function with respect to the model parameters is computed based on the selected sample.
+- The model parameters are updated using the computed gradient and the learning rate.
+- The process is repeated for multiple iterations until convergence or a specified number of epochs.
+
+## Formula 
+
+$$ 
+\large \theta = \theta - \alpha  * \nabla J((\theta ; x_iy_i))
+$$
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:37]:
```suggestion
$$ 
\large \theta = \theta - \alpha  \cdot  \nabla J(\theta ; x_i, y_i)
$$
``` 
Typically \cdot is used for multiplication to keep the LaTex notation standard and  the ` * ` symbol is not used.

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+- θ represents the model parameter (weight or bias) being updated.
+- α is the learning rate, a hyperparameter that controls the step size of the update.
+- ∇J(θ;xi,yi)  is the gradient of the cost or loss function J with respect to the model parameter θ, computed based on a single training sample (xi,yi).
+
+## Advantages
+- **Faster convergence:** SGD updates parameters more frequently hence it takes less time to converge especially for large datasets.
+- **Reduced Computation Time:** SDD takes only subset of dataset or batch for each update. This makes it easy to handle large datasets and compute faster.
+- **Avoid Local Minima:** The noise introduced by updating parameters with individual data points or small batches can help escape local minima.This can potentially lead to better solutions in complex, non-convex optimization problems.
+- **Online Learning:** SGD can be used in scenarios where data is arriving sequentially (online learning).- It allows models to be updated continuously as new data comes in.
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:49]:
```suggestion
## Advantages
- **Faster convergence:** SGD updates parameters more frequently hence it takes less time to converge especially for large datasets.
- **Reduced Computation Time:** SGD takes only a subset of dataset or batch for each update. This makes it easy to handle large datasets and compute faster.
- **Avoid Local Minima:** The noise introduced by updating parameters with individual data points or small batches can help escape local minima.This can potentially lead to better solutions in complex, non-convex optimization problems.
- **Online Learning:** SGD can be used in scenarios where data is arriving sequentially (online learning).- It allows models to be updated continuously as new data comes in.
```

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+## Disadvantages
+- **Noisy Updates:** Updates are based on a single data point or small batch, which introduces variability in the gradient estimates.This noise can cause the algorithm to converge more slowly or oscillate around the optimal solution.
+- **Convergence Issues:** The noisy updates can lead to less stable convergence and might make it harder to reach the exact minimum of the loss function.Fine-tuning the learning rate and other hyperparameters becomes crucial to achieving good results.
+- **Hyperparameter Sensitivity:** - SGD's performance is sensitive to the choice of learning rate and other hyperparameters.Finding the right set of hyperparameters often requires experimentation and tuning.
+
+## Practical Tips And Tricks When Using SGD
+- Shuffle data before training 
+- Use mini batches(batch size 32)
+- Normalize input
+- Choose suitable learning rate (0.01)
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:60]:
```suggestion
## Practical Tips And Tricks When Using SGD
- Shuffle data before training 
- Use mini batches(batch size 32)
- Normalize input
- Choose a suitable learning rate (0.01)
```

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @pratik305 ,
I had a look at your file and suggested some changes to you please make them asap. Also  include an example section and an syntax section as per the [issue](https://github.com/Codecademy/docs/issues/4527) related to this says and also include your file in the correct path.

**Comment by pratik305** [comment]
@SaviDahegaonkar  all changes are done

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  - 'Computer Science'
+Tags:
+  - 'AI'
+  - 'Neural Network'
+CatalogContent:
+  - 'paths/computer-science'
+  - 'paths/data-science'
+---
+
+**Stochastic Gradient Descent** (SGD) is an optimization algorithm. It is variant of gradient descent optimizer. The SGD minimizes the loss function of machine learning algorithms and deep learning algorithms during backpropagation to update the weights and biases in Artificial Neural Networks.
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:15]:
```suggestion
**Stochastic Gradient Descent** (SGD) is an optimization algorithm. It is a variant of gradient descent optimizer. The SGD minimizes the loss function of machine learning algorithms and deep learning algorithms during backpropagation to update the weights and biases in Artificial Neural Networks. 
```

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+
+![stochastic gradient descent](https://www.goglides.dev/images/Jq8EpuPoMjCcxm7PqMqWuQK7M_MoVtdfAUsGJsoUIMA/w:880/mb:500000/ar:1/aHR0cHM6Ly93d3ct/Z29nbGlkZXMtZGV2/LnMzLmFtYXpvbmF3/cy5jb20vdXBsb2Fk/cy9hcnRpY2xlcy8z/cGh3bjR0bmpnNGlo/eHV0Znpqby5wbmc)
+
+## Algorithms Step
+
+- At each iteration, a random sample  is selected from the training dataset.
+- The gradient of the cost function with respect to the model parameters is computed based on the selected sample.
+- The model parameters are updated using the computed gradient and the learning rate.
+- The process is repeated for multiple iterations until convergence or a specified number of epochs.
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:29]:
```suggestion
- At each iteration, a random sample is selected from the training dataset.
- The gradient of the cost function with respect to the model parameters is computed based on the selected sample.
- The model parameters are updated using the computed gradient and the learning rate.
- The process is repeated for multiple iterations until convergence or a specified number of epochs.
```

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+- Shuffle data before training 
+- Use mini batches(batch size 32)
+- Normalize input
+- Choose a suitable learning rate (0.01)
+
+## Syntax
+- Learning Rate (α): A hyperparameter that controls the size of the update step.
+- Number of Iterations: The number of times the algorithm will iterate over the dataset.
+- Loss Function: The function that measures the error of the model predictions.
+- Gradient Calculation: The method for computing gradients based on the loss function.
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:64]:
```suggestion
## Syntax
- Learning Rate (α): A hyperparameter that controls the size of the update step.
- Number of Iterations: The number of times the algorithm will iterate over the dataset.
- Loss Function: The function that measures the error of the model predictions.
- Gradient Calculation: The method for computing gradients based on the loss function.
```
The syntax must be included inside the backticks (```.....```) in the pseudo block. I can't find any syntax here. Please include a syntax.

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @pratik305 ,
I have suggested a few grammatical corrections and also included a relevant syntax and example section.

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md (hunk 9 above)
Comment (pratik305) [reply content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:64]: @SaviDahegaonkar  there is no syntax for sgd. we can use sgd with tensorflow, pytorch, numpy,scikit-learn each one of them are different syntax. what to do

Comment (pratik305) [reply content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:64]: can i add syntax like this 
`SGD(learning_rate, n_iterations, loss_function, gradient_calculation)`

and then add code like 
def stochastic_gradient_descent(X, y, theta, learning_rate, n_iterations):
    for iteration in range(n_iterations):
        for i in range(len(y)):
            gradient = compute_gradient(X[i], y[i], theta)
            theta -= learning_rate * gradient
    return theta

**Comment by pratik305** [comment]
@SaviDahegaonkar  sir reply fast that can i make changes

**Comment by SaviDahegaonkar** [comment]
> @SaviDahegaonkar sir reply fast that can i make changes

Yes sure you can, I will approve this PR.

**Comment by pratik305** [comment]
@SaviDahegaonkar all done

File: content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+## Example
+```python
+  def stochastic_gradient_descent(X, y, theta, learning_rate, n_iterations):
+      for iteration in range(n_iterations):
+      for i in range(len(y)):
+      gradient = compute_gradient(X[i], y[i], theta)
+      theta -= learning_rate * gradient
+      return theta
```
Comment (SaviDahegaonkar) [inline content/ai/concepts/neural-networks/terms/stochastic-gradient-descent/stochastic-gradient-descent.md:76]:
```suggestion
  def stochastic_gradient_descent(X, y, theta, learning_rate, n_iterations):
      for iteration in range(n_iterations):
        for i in range(len(y)):
          gradient = compute_gradient(X[i], y[i], theta)
          theta -= learning_rate * gradient
      return theta
```
Added some proper indentation.

**Review by SaviDahegaonkar — APPROVED** [review]
Hey @pratik305 ,
LGTM!

Thanks,
Savi

**Comment by pratik305** [comment]
@avdhoottt all changes are done can you review this please

**Comment by pratik305** [comment]
@avdhoottt when will it get merged

**Comment by pratik305** [comment]
@avdhoottt When will this is get this merge. Tell is there any think to change

**Comment by pratik305** [comment]
@Maheshwaran17 can you please check this

**Comment by pratik305** [comment]
@avdhoottt  when will it get merge sir  is there any problem

**Review by avdhoottt — APPROVED** [review]
LGTM!

**Comment by avdhoottt** [comment]
> @avdhoottt when will it get merge sir is there any problem

Hey @pratik305, sorry for the late reply. I'm merging this PR. Thank you so much for contributing!
